C error: conflicting types when calling a function in an anonymous function

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relvant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
**V version:** 0.2.1 204cc5f.59e5e32
**OS:** Linux version 5.10.6-arch1-1

**What did you do?**
Calling a function that returns a value (without seems to work fine) from inside an anonymous function, that is declared after the the anonymous function.

```v
module main

fn main() {
	anon_fn := fn() {
		println(test())
	}
	anon_fn()
}

fn test() string {
	return "Test"
}
```

**What did you expect to see?**
Successful compilation and running (printing "Test")
 
**What did you see instead?**
```
==================
      |           ^~~~~~~~~~
/tmp/v/test.2667229229384582287.tmp.c:1389:11: error: incompatible type for argument 1 of ‘println’
 1389 |   println(main__test());
      |           ^~~~~~~~~~~~
      |           |
      |           int
/tmp/v/test.2667229229384582287.tmp.c:1089:21: note: expected ‘string’ but argument is of type ‘int’
 1089 | void println(string s);
      |              ~~~~~~~^
/tmp/v/test.2667229229384582287.tmp.c: At top level:
/tmp/v/test.2667229229384582287.tmp.c:1392:24: error: conflicting types for ‘main__test’
 1392 | VV_LOCAL_SYMBOL string main__test();
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.
```

Moving the `test` function in front of `main` fixes the problem. :thinking:

--- a/vlib/v/gen/cgen.v
+++ b/vlib/v/gen/cgen.v
@@ -93,6 +93,7 @@ mut:
 	map_fn_definitions    []string // map equality functions that have been defined
 	struct_fn_definitions []string // struct equality functions that have been defined
 	auto_fn_definitions   []string // auto generated functions defination list
+	anon_fn_definitions   []string // anon generated functions defination list
 	is_json_fn            bool     // inside json.encode()
 	json_types            []string // to avoid json gen duplicates
 	pcs                   []ProfileCounterMeta // -prof profile counter fn_names => fn counter name
@@ -338,6 +339,11 @@ pub fn cgen(files []ast.File, table &table.Table, pref &pref.Preferences) string
 	}
 	if g.auto_fn_definitions.len > 0 {
 		for fn_def in g.auto_fn_definitions {
+			b.writeln(fn_def)
+		}
+	}
+	if g.anon_fn_definitions.len > 0 {
+		for fn_def in g.anon_fn_definitions {
 			b.writeln(fn_def)
 		}
 	}
@@ -2361,12 +2367,10 @@ fn (mut g Gen) autofree_var_call(free_fn_name string, v ast.Var) {
 
 fn (mut g Gen) gen_anon_fn_decl(it ast.AnonFn) {
 	pos := g.out.len
-	def_pos := g.definitions.len
 	g.stmt(it.decl)
 	fn_body := g.out.after(pos)
 	g.out.go_back(fn_body.len)
-	g.definitions.go_back(g.definitions.len - def_pos)
-	g.definitions.write(fn_body)
+	g.anon_fn_definitions << fn_body
 }
 
 fn (mut g Gen) map_fn_ptrs(key_typ table.TypeSymbol) (string, string, string, string) {

--- a/vlib/v/tests/anon_fn_call_test.v
+++ b/vlib/v/tests/anon_fn_call_test.v
@@ -1,0 +1,10 @@
+fn test_anon_fn_call() {
+	anon_fn := fn() string {
+		return test()
+	}
+	assert anon_fn() == 'Test'
+}
+
+fn test() string {
+	return "Test"
+}

